### ISSUE TITLE
emacs-lisp checker doesn't report compilation errors caused by unbalanced parenthesis

### ISSUE THREAD

**Issue by pkryger** [flycheck/flycheck#2092]
## Checklist

- [x] I have checked existing issues for potential duplicates before creating this one.
- [x] I have read the [Troubleshooting guide][].

## Bug description

When parenthesis are unbalanced in a file, `flycheck-elisp` report no errors, even though they are visible in `flycheck-compile` output. The two distinct errors I have observed as not reported are:
1. When there are to few closing parenthesis: `Error: End of file during parsing`
2. When there are too many closing parenthesis: `Error: Invalid read syntax: ")", 13, 1`

This issue, in some cases, causes mode-line lighter and `*Flycheck errors*` buffer to report no errors whatsoever, even though the file is not a valid lisp in the first place.

## Steps to reproduce

1. On vanilla Emacs open file a new file:
```bash
$ emacs ~/tmp/foo.el
```
3. Insert
```emacs-lisp
;; -*- lexical-binding: t -*-

(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
(package-initialize)
(package-refresh-contents)
(package-install 'flycheck)
(declare-function flycheck-mode "flycheck")
(defvar flycheck-disabled-checkers)
(add-to-list 'flycheck-disabled-checkers 'emacs-lisp-checkdoc)
(flycheck-mode)
```
4. Do `M-x eval-buffer`.
5. Alternatively:
  - to reproduce error 1: delete last parenthesis
  - to reproduce error 2: insert extra parenthesis
6. Do `C-c ! C-c RET` to see compilation output, to confirm the errors are reported.

- [ ] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## Expected behavior

Both errors are reported in mode-line lighter`*Flycheck errors*`

## Screenshots

<img width="1024" alt="Screenshot 2024-11-08 at 12 44 45" src="https://github.com/user-attachments/assets/b1f278b6-5b96-4872-abda-d778b7b5f220">
<img width="1020" alt="Screenshot 2024-11-08 at 12 44 23" src="https://github.com/user-attachments/assets/01d1ccbe-decb-4636-8b2a-ebb6d51dc975">

## System configuration
On macOS 15.0.1 Sequoia.
- Emacs-29.4: [emacs-mac](https://bitbucket.org/mituharu/emacs-mac/commits/branch/master), self compiled from [railwaycat/homebrew-emacsmacport](https://github.com/railwaycat/homebrew-emacsmacport)
- Emacs-31.0.50: [nix-emacs-ci](https://github.com/purcell/nix-emacs-ci) self build from `master` tag from [Silex/docker-emacs](https://github.com/Silex/docker-emacs)

```
Syntax checkers for buffer foo.el in emacs-lisp-mode:

First checker to run:

  emacs-lisp
    - may enable:    yes
    - may run:       t
    - executable:    Found at /nix/store/emacs/bin/emacs
    - next checkers: emacs-lisp-checkdoc

Checkers that are compatible with this mode, but will not run until properly configured:

  emacs-lisp-checkdoc (manually disabled) enable
    - may enable: yes
    - executable: Found at /nix/store/emacs/bin/emacs

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 20240726.456
Emacs version:    31.0.50
System:           aarch64-unknown-linux-gnu
Window system:    nil
```

Emacs configuration:
- [x] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [x] Other shared configuration: [Exordium](https://github.com/emacs-exordium/exordium)

## Additional notes

As a workaround I added extra error pattern to `(flycheck-define-checker emacs-lisp [...]`:
```
   (error line-start (file-name) ":"
          (zero-or-more whitespace) "Error:" (zero-or-more whitespace)
          (message (or "End of file during parsing"
                       (seq "Invalid read syntax:" (zero-or-more (not ",")))))
          (optional "," (zero-or-more whitespace) line
                    "," (zero-or-more whitespace) column)
```
Happy to cut a PR if the above solution has a potential of becoming a permanent fix.

[Troubleshooting guide]:  https://www.flycheck.org/en/latest/user/troubleshooting.html

**Comment by pkryger** [comment]
FWIW: I made a slight improvement to my workflow (where I reinstall packages fairly often), by adding the following to my config:
```emacs-lisp
(use-package flycheck
  :autoload (flycheck-rx-to-string
             flycheck-checker-get)
  :config
  ;; Fix for https://github.com/flycheck/flycheck/issues/2092
  (let* ((p '(error line-start (file-name) ":"
                   (zero-or-more whitespace) "Error:" (zero-or-more whitespace)
                   (message (or "End of file during parsing"
                                (seq "Invalid read syntax:" (zero-or-more (not ",")))))
                   (optional "," (zero-or-more whitespace) line
                             "," (zero-or-more whitespace) column)))
         ;; pre-compile pattern like `flycheck-define-command-checker' does
         (pattern (cons (flycheck-rx-to-string `(and ,@(cdr p))
                                               'no-group)
                        (car p))))
    (unless (cl-find-if (lambda (p)
                          (and (eq (cdr p) (cdr pattern))
                               (equal (car p) (car pattern))))
                        (flycheck-checker-get 'emacs-lisp 'error-patterns))
      (push pattern (flycheck-checker-get 'emacs-lisp 'error-patterns))))
```

**Comment by bbatsov** [comment]
This has been fixed by adding error patterns for "End of file during parsing" and "Invalid read syntax" errors to the emacs-lisp checker, so compilation errors caused by unbalanced parentheses are now properly reported.
